### ISSUE TITLE
Event content recreated on each rerender

### ISSUE THREAD

**Issue by Aleafan** [fullcalendar/fullcalendar#7650]
### Reduced Test Case

https://stackblitz.com/github/fullcalendar/fullcalendar-examples/tree/main/react

### Do you understand that if a reduced test case is not provided, we will intentionally delay triaging of your ticket?

- [X] I understand

### Which connector are you using (React/Angular/etc)?

React

### Bug Description

Hello, thanks a lot for a good job!

I use FullCalendar in React app and render custom event content. I found that event content wrapper is recreated anew each rerender vs simple rerender. This causes all event content to also be dismounted and mounted each time. And this caused some subtle bugs like loss of focus from inside event content and also some performance penalty.

I found that in e.g. daygrid view the root of the problem is that key generation function generateSegKey() produces new key each rerender hence React just destroys this whole subtree and recreates it anew.
```
// From TableRow.tsx file
nodes.push(
          <div
            className={'fc-daygrid-event-harness' + (isAbsolute ? ' fc-daygrid-event-harness-abs' : '')}
            key={generateSegKey(seg)}
            ...
```
And new keys are produced due to each event instanceId is changing every rerender.

Is that a real problem or intentional design?

**Comment by SHergibo** [comment]
Hi @Aleafan, is your problem the same as this one?

https://github.com/fullcalendar/fullcalendar/issues/7488

**Comment by Aleafan** [comment]
> Hi @Aleafan, is your problem the same as this one?
> 
> #7488

Hi @SHergibo, I don't really know. I have React 17.0.2 and no events flickering, but it could have the same root cause.

Looking at your case I thought that maybe you could try to wrap your event generating function in useCallback so it has stable reference between renders. I found that if this function is recreated it will be called unnecesserily on every render instead of how its described in docs.

And I can share what I changed in FC code for daygrid and timegrid views for my problem to disappear. Those are files for patch-package library, can be used directly with it or just read as diff between old/fixed version.

[@fullcalendar+daygrid+6.1.10.patch](https://github.com/fullcalendar/fullcalendar/files/14934082/%40fullcalendar%2Bdaygrid%2B6.1.10.patch)
[@fullcalendar+timegrid+6.1.10.patch](https://github.com/fullcalendar/fullcalendar/files/14934084/%40fullcalendar%2Btimegrid%2B6.1.10.patch)

**Comment by acerix** [comment]
Yes, the issue seems caused by re-initialising the calendar on state changes so I believe it's the same as that existing ticket :
https://github.com/fullcalendar/fullcalendar/issues/7488
↑
Please visit the above issue, **give a** 👍  **to vote** ([?](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments)), press the 🔈**Subscribe** button to receive updates.

**Comment by arshaw** [comment]
Reopening this issue because it's distinct from #7488. This issue is related to the React `key` on `fc-daygrid-event-harness` being frequently invalidated. @Aleafan, I began addressing this in the v7 beta yesterday.

**Comment by arshaw** [comment]
I'm actually questioning this now. @Aleafan, you say `instanceId` is changing every render... but `instanceId` would only change if a distinct new event object is given, thus resulting in new rendered event. Can you provide a better recreation? The on you provided seems to be the unaltered stock repro

**Comment by Aleafan** [comment]
Hello @arshaw, sorry for late response! I tried to recreate the problem on my project, but our API has changed quite a lot and now I couldn't get back to older setup. So I'll try to just describe it.

Our setup had events coming from the server and passed to Calendar via fuction. We had a custom eventContent injection with React elements passed to extendedProps.children for every event. On click to event a popup was opening (passed to updated children of corresponding event). This popup was immediately closed because fc-daygrid-event-harness changed its key, destroyed dom and lost focus that triggered popup closing. Key changing happened because event's instanceId was changed when in reality it was the same event with some added data.

I don't know how instanceId is formed but in our current setup its stable and I can remove my patches with no problems. I remember though that back then events had new instanceId every rerender of parent component event when nothing changed.

So do you see the setup as described where user changes event but expects it to have focus after minor change? In that case it could be good to remove key from wrapper, spread list and allow React to compare elements without keys help. Otherwise if instanceId logic is robust this point can be closed.

**Comment by arshaw** [comment]
This has been improved in [v7.0.0-rc.0](https://github.com/fullcalendar/fullcalendar-workspace/releases/tag/v7.0.0-rc.0), please see [this comment in another issue](https://github.com/fullcalendar/fullcalendar/issues/3003) for more details.

**Comment by gnapse** [comment]
> see [this comment in another issue](https://github.com/fullcalendar/fullcalendar/issues/3003) for more details.

the link does not point to a comment in that other issue, but to the issue itself. Can you clarify what comment you refer to?
